src/distancematrix.rs

```diff
@@ -0,0 +1,220 @@
+///
+/// let mut t = DistanceMatrix::new(vec!["a".to_string(), "b".to_string()],
+/// vec![
+///  vec![],
+///  vec![1.0],
+/// ]).unwrap();
+/// assert_eq!(t[(1, 0)], 1.0);
+/// t[(0, 1)] = 2.0;
+/// assert_eq!(t[(1, 0)], 2.0);
+/// ```
+impl Index<(usize, usize)> for DistanceMatrix {
+    type Output = f32;
+    fn index(&self, t: (usize, usize)) -> &Self::Output {
+        let (i, j) = self.get_index(t);
+        &self.distances[i][j]
+    }
+}
+
+/// Mutable index access into the DistanceMatrix, taking into account the shape, like `Index`.
+impl IndexMut<(usize, usize)> for DistanceMatrix {
+    fn index_mut(&mut self, t: (usize, usize)) -> &mut Self::Output {
+        let (i, j) = self.get_index(t);
+        &mut self.distances[i][j]
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn indexing() {
+        let mut m = DistanceMatrix::new(
+            vec!["a".to_string(), "b".to_string(), "c".to_string()],
+            vec![vec![0., 1., 2.], vec![3., 4., 5.], vec![6., 7., 8.]],
+        )
+        .unwrap();
+        assert_eq!(m.matrix_type, MatrixType::Square);
+        assert_eq!(m[(0, 0)], 0.);
+        assert_eq!(m[(1, 2)], 5.);
+        assert_eq!(m[(2, 1)], 7.);
+        m[(1, 2)] = -1.;
+        m[(2, 1)] = -2.;
+        assert_eq!(m[(1, 2)], -1.);
+        assert_eq!(m[(2, 1)], -2.);
+    }
+
+    #[test]
+    fn lower_indexing() {
+        let mut m = DistanceMatrix::new(
+            vec!["a".to_string(), "b".to_string(), "c".to_string()],
+            vec![vec![], vec![1.], vec![2., 3.]],
+        )
+        .unwrap();
+        assert_eq!(m.matrix_type, MatrixType::Lower);
+        assert_eq!(m[(2, 0)], 2.);
+        assert_eq!(m[(0, 2)], 2.);
+        assert_eq!(m[(1, 2)], 3.);
+        assert_eq!(m[(2, 1)], 3.);
+        m[(0, 2)] = -1.;
+        m[(2, 1)] = -2.;
+        m[(1, 2)] = -3.;
+        assert_eq!(m[(0, 2)], -1.);
+        assert_eq!(m[(2, 0)], -1.);
+        assert_eq!(m[(1, 2)], -3.);
+        assert_eq!(m[(2, 1)], -3.);
+    }
+
+    #[test]
+    fn upper_indexing() {
+        let mut m = DistanceMatrix::new(
+            vec!["a".to_string(), "b".to_string(), "c".to_string()],
+            vec![vec![1., 2.], vec![3.], vec![]],
+        )
+        .unwrap();
+        assert_eq!(m.matrix_type, MatrixType::Upper);
+        assert_eq!(m[(0, 2)], 2.);
+        assert_eq!(m[(1, 2)], 3.);
+        assert_eq!(m[(2, 1)], 3.);
+        m[(0, 2)] = -1.;
+        m[(2, 1)] = -2.;
+        m[(1, 2)] = -3.;
+        assert_eq!(m[(0, 2)], -1.);
+        assert_eq!(m[(2, 0)], -1.);
+        assert_eq!(m[(1, 2)], -3.);
+        assert_eq!(m[(2, 1)], -3.);
+    }
+
+    #[test]
+    fn bad_length() {
+        let m = DistanceMatrix::new(
+            vec!["a".to_string(), "b".to_string(), "c".to_string()],
+            vec![vec![1., 2.], vec![3.]],
+        );
+        match m {
+            Err(DistanceMatrixError::LengthError) => (),
+            _ => assert!(false),
```

Comment:
There is a `assert_match!` macro in the `assert_matches` package to make this cleaner, but that would be a new dependency

Reply:
Ah, `assert!(matches!(..))` is supported without packages :)